### PR TITLE
Fix virtual to real path mapping in Plugin Translator on dmtcp_restart using DMTCP_PATH_MAPPING

### PR DESCRIPTION
**Pull request by ishanaro4**
**Problem Description**

This PR addresses an issue previously reported in https://github.com/dmtcp/dmtcp/issues/1232.
When using DMTCP_PATH_MAPPING to relocate checkpoint files to a new directory, repeated path translations result in incorrect path expansion.

**Scenario**

Suppose I have my application executable and related files in /a/b/c, So at the time of checkpoint, it would save this path to all the files inside /a/b/c.
Now I copies contents inside "c" folder including checkpoint files to new location , suppose /q/w/e. Now if I delete the older location and try to run dmtcp_restart with DMTCP_PATH_MAPPING variable i.e below command 
`DMTCP_PATH_MAPPING="/a/b/c:/q/w/e" dmtcp_restart ckpt_*`

**Observed Behaviour**

The first path translation works correctly: `/a/b/c → /q/w/e`.
However, subsequent translations of the same virtual path result in incorrect path construction:
`/a/b/c → /q/w/ee
→ /q/w/eee
→ /q/w/eeee...`
This eventually causes:

> - Invalid paths
> - File not found errors
> - dmtcp hanging during restart

**Root Cause**

The issue appears to originate from the following line:
`int n = snprintf(data->virtualToRealPath.path, PATH_MAX, "%s%s", newp.c_str(), suffix);`
Repeated translations append the suffix multiple times, causing cumulative path corruption.
Additionally, there was no guard preventing already-translated paths from being reprocessed.

**Fix**

1. Corrected the snprintf usage to avoid unintended concatenation.
2. Added a guard condition to prevent re-translation of paths that already contain the mapped prefix.
`!(Util::strStartsWith(virtPath, newp.c_str()))`

**Testing**

> - Reproduced the issue using the scenario described above.
> - Verified correct behavior after fix:
> - Path translation happens only once.
> - No cumulative suffix growth.
> - Restart completes successfully.
Please let me know if additional test cases or automated tests should be added.

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed an edge case in path translation that could produce incorrect translated paths in certain overlaps; preserves existing boundary checks and path-length validation.
  * No changes to public interfaces or exported signatures; behavior and validations remain unchanged.

[Review Change Stack](https://app.coderabbit.ai/change-stack/dmtcp/dmtcp/pull/1238)

### CODE REVIEW

**Comment by karya0** [comment]
> The first path translation works correctly: `/a/b/c → /q/w/e`. However, subsequent translations of the same virtual path result in incorrect path construction: `/a/b/c → /q/w/ee → /q/w/eee → /q/w/eeee...` This eventually causes:

I a not sure what do you mean by subsequent translations. Since the mapping has only one key-value pair, the for loop should only be executed once. 

I think the root cause (as you have also hinted at) is that we are overwriting the original string. If we are translating with mapping "/a/b/c:/q/w/sss" and the input string is "/a/b/c/d.txt",  snprintf will never terminate because we have suffix pointing to "/d.txt". So in a way, the source and destination overlaps will cause infinite loop. The right fix here would be to make a copy of suffix before snprintf is called. I'll add a suggestion.

File: src/plugin_pathtranslator.cpp [hunk 1]
```diff
@@ -137,12 +137,16 @@ pathTranslator_VirtualToReal(DmtcpEventData_t *data)
...
       char next = virtPath[oldp.size()];
       if (next == '/' || next == '\0') {
-        const char* suffix = virtPath + oldp.size();
-        int n = snprintf(data->virtualToRealPath.path, PATH_MAX, "%s%s", newp.c_str(), suffix);
+        string original = virtPath;  
+        string replaced = newp + original.substr(oldp.size());
+        int n = snprintf(data->virtualToRealPath.path,
+                         PATH_MAX,
+                         "%s",
+                         replaced.c_str());
```
Comment (karya0) [inline src/plugin_pathtranslator.cpp]:
```suggestion
        string suffix = &virtPath[oldp.size()];  
        int n = snprintf(data->virtualToRealPath.path,
                         PATH_MAX,
                         "%s%s",
                         newp.c_str(),
                         suffix.c_str());
```

**Review by karya0 — DISMISSED** [review]
Please let me know if this fix would work.

**Review by gc00 — APPROVED** [review]
LGTM.  Updating for new dot release.

**Review by gc00 — APPROVED** [review]
LGTM
